utils/IrodsConnector.py

```diff
@@ -973,10 +983,13 @@ def upload_data(self, src_path, dst_coll, resc_name, size, buff=BUFF_SIZE,
                 diff, only_fs, _, _ = self.diffObjFile(
                     dst_path, src_path, scope='checksum')
             else:
+                compare_coll = self.ensure_coll(
+                          dst_coll.path+'/'+os.path.basename(src_path))
                 diff, only_fs, _, _ = self.diffIrodsLocalfs(
-                    dst_coll, src_path)
+                    compare_coll, src_path)
```

Comment:
`dst_path` already contains the path of `compare_coll` here.  The **bug** is that the original `dst_coll` is used instead of a collection created from `dst_path`.  So, just create the new collection with a more appropriate name for the current convention:

```python
# line 972
                cmp_path = utils.utils.IrodsPath(dst_coll.path, src_path.name)
```
```python
# line 983
                diff, only_fs, _, _ = self.diffObjFile(
                    cmp_path, src_path, scope='checksum')
            else:
                cmp_coll = self.ensure_coll(cmp_path)
                diff, only_fs, _, _ = self.diffIrodsLocalfs(
                    cmp_coll, src_path)
```

Reply:
Where the code is confusing, I do agree with you. I cannot wait until this merge is done and we can break these big functions up into small logical pieces.